pkg/ipam-node/migrator/migrator.go

```diff
@@ -0,0 +1,159 @@
+	storePkg "github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-node/store"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-node/types"
+)
+
+const (
+	// EnvDisableMigration contains the name of the environment variable which can be used
+	// to disable migration
+	EnvDisableMigration = "MIGRATOR_DISABLE_MIGRATION"
+	// EnvHostLocalStorePath contains the name of the environment variable which can be used to
+	// change host local IPAM store path
+	EnvHostLocalStorePath = "MIGRATOR_HOST_LOCAL_STORE"
+	// DefaultHostLocalStorePath contains default path for host-local store which was used
+	// in the older version
+	DefaultHostLocalStorePath = "/var/lib/cni/nv-ipam/state/host-local"
+	// PlaceholderForUnknownField contains placeholder string which will be used as a value
+	// for fields for which we have no data
+	PlaceholderForUnknownField = "MIGRATOR_NO_DATA"
+)
+
+// New create and initialize new instance of the migrator
+func New(store storePkg.Store) Migrator {
+	return &migrator{
+		store: store,
+	}
+}
+
+// Migrator is the interface implemented by migrator package
+type Migrator interface {
+	// Migrate execute migration logic.
+	// The implementation will check if host-local IPAM data that were created by the previous version
+	// of nv-ipam are available on the filesystem. If the data is found,
+	// the migrator will read and save them into the store and then remove data in the old format.
+	// Some information required by the new store schema can't be restored from the host-local IPAM store
+	// format used in the older version of the nv-ipam. Missing data will be replaced by
+	// a special placeholder which indicates that data is missing.
+	Migrate(ctx context.Context) error
+}
+
+// default migrator implementation
+type migrator struct {
+	store storePkg.Store
+}
+
+// Migrate is the Migrator interface implementation for the migrator
+func (m *migrator) Migrate(ctx context.Context) error {
+	logger := logr.FromContextOrDiscard(ctx).WithName("migrator")
+	if os.Getenv(EnvDisableMigration) != "" {
+		logger.Info(fmt.Sprintf("%s set, skip migration", EnvDisableMigration))
+		return nil
+	}
+	hostLocalPath := os.Getenv(EnvHostLocalStorePath)
+	if hostLocalPath == "" {
+		hostLocalPath = DefaultHostLocalStorePath
+	}
+	info, err := os.Stat(hostLocalPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			logger.Info("host-local IPAM path not found, skip migration", "path", hostLocalPath)
+			return nil
+		}
+		return err
+	}
+	if !info.IsDir() {
+		return fmt.Errorf("host-local IPAM path is not a dir")
+	}
+	session, err := m.store.Open(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to open store: %v", err)
+	}
+	logger.Info("check host-local store path", "path", hostLocalPath)
+	err = filepath.Walk(hostLocalPath, getWalkFunc(logger, session))
+	if err != nil {
+		session.Cancel()
+		return err
+	}
+	err = session.Commit()
+	if err != nil {
+		return err
+	}
+	err = os.RemoveAll(hostLocalPath)
+	if err != nil {
```

Comment:
if removal fails, then subsequent migrations attempts will fail.

should you try to remove added entries from store so that next round might succeed ?

Reply:
alternative is to keep as is, let user delete the dir on affected nodes as its unlikely we will hit this issue

Reply:
> should you try to remove added entries from store so that next round might succeed ?

During the migration loop we ignore "Reservation errors", so 2 attempt should succeed if the condition which leads to os.Remove to fail gone.